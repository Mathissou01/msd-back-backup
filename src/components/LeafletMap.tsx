import { GeoJSON, Polygon, Map } from "leaflet";
import React, { useRef } from "react";
import { FeatureGroup, MapContainer, TileLayer } from "react-leaflet";
import { EditControl } from "react-leaflet-draw";
import "leaflet/dist/leaflet.css";
import "leaflet-draw/dist/leaflet.draw.css";
import "leaflet-defaulticon-compatibility/dist/leaflet-defaulticon-compatibility.css";
import "leaflet-defaulticon-compatibility";
import { fetchAzureSearch } from "../lib/api";
import styles from "../styles/leaflet-map.module.scss";

interface IPolygon {
  type: string;
  properties: {};
  geometry: {
    type: string;
    coordinates: Array<Array<Array<Number>>>;
  };
}

interface IGeoResult {
  cle_interop: string;
  commune_insee: string;
  commune_nom: string;
  voie_nom: string;
  numero: string;
}

const LeafletMap = () => {
  let layerRef = useRef(null);
  const [geoResults, setGeoResults] = React.useState<Array<IGeoResult>>([]);

  async function toGeoJSON() {
    let collection: GeoJSON.FeatureCollection = {
      type: "FeatureCollection",
      features: [],
    };
    let layerGroup: Map = layerRef.current as unknown as Map;
    layerGroup.eachLayer((layer) => {
      if (layer instanceof Polygon) {
        collection.features.push(layer.toGeoJSON());
      }
    });

    if (collection.features.length > 0) {
      console.log(collection);
      let formattedCollection: Array<string> = [];
      collection.features.forEach((feature) => {
        const polygon = feature as IPolygon;
        const formattedCoords = polygon.geometry.coordinates[0].map(
          (coordinate) => {
            return `${coordinate[0]} ${coordinate[1]}`;
          },
        );
        formattedCollection.push(
          `geo.intersects(coordonnees, geography'POLYGON((${formattedCoords}))')`,
        );
      });

      await fetchAzureSearch("/indexes/adresses-gouv/docs", {
        "api-version": "2020-06-30",
        $count: true,
        $orderby: "voie_nom,numero",
        $select: "cle_interop,commune_insee,commune_nom,voie_nom,numero",
        $filter: formattedCollection.join(" or "),
      }).then((results) => {
        // console.log("RESULTS", results);
        // console.log("COUNT", results["@odata.count"]);
        setGeoResults(results.value);
      });
    } else {
      setGeoResults([]);
    }
  }

  return (
    <>
      <MapContainer
        center={[45.777, 4.85973]}
        zoom={19}
        scrollWheelZoom={true}
        style={{ height: "100%", width: "100%" }}
      >
        <TileLayer
          url="https://{s}.tile.openstreetmap.fr/osmfr/{z}/{x}/{y}.png"
          attribution='&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'
        />
        <FeatureGroup ref={layerRef}>
          <EditControl
            position={"topright"}
            draw={{
              rectangle: false,
            }}
          />
        </FeatureGroup>
      </MapContainer>
      <button onClick={toGeoJSON} className={styles.leafletMap__button}>
        Récupérer Adresses
      </button>
      <div className={styles.leafletMap__results}>
        Résultats ({geoResults.length}) :
      </div>
      {geoResults?.length > 0 && (
        <ul>
          {geoResults?.map((result: IGeoResult) => {
            return (
              <li key={result.cle_interop}>
                <p>
                  {result.numero +
                    " " +
                    result.voie_nom +
                    ", " +
                    result.commune_insee +
                    " " +
                    result.commune_nom}
                </p>
              </li>
            );
          })}
        </ul>
      )}
    </>
  );
};

export default LeafletMap;
