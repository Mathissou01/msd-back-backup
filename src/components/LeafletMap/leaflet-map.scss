@use "../../styles/06-vendors/vendors.leaflet";

.c-LeafletMap {
  &,
  > div {
    width: 100%;
    height: 100%;
  }

  &__Button {
    margin: 20px;
  }

  &__Results {
    margin: 0 20px;
  }
}
