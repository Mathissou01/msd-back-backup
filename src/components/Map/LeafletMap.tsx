import React from "react";
import { Wrapper, Status } from "@googlemaps/react-wrapper";
import { MapContainer, TileLayer } from "react-leaflet";
import L, { CRS } from "leaflet";
import EditableGroup from "./FeatureGroup/EditableGroup";
import CommonSpinner from "../Common/CommonSpinner/CommonSpinner";
import { ISectorPolygon } from "../../lib/sectors";
import "leaflet/dist/leaflet.css";
import "leaflet-draw/dist/leaflet.draw.css";
import "./leaflet-map.scss";

//TODO: Select one polygon to edit (one polygon at a time)
// function EditableLayer(props) {
//   const editLayerRef = React.useRef();
//   let drawControlRef = React.useRef();
//   const map = useMap();
//   console.log("hi", drawControlRef);
//   useEffect(() => {
//     if (!props.showDrawControl) {
//       map.removeControl(drawControlRef.current);
//     } else {
//       map.addControl(drawControlRef.current);
//     }
//     editLayerRef.current.clearLayers();
//     editLayerRef.current.addLayer(props.layer);
//     props.layer.on("click", function (e) {
//       props.onLayerClicked(e, drawControlRef.current);
//     });
//   }, [props, map]);
//   function onMounted(ctl) {
//     drawControlRef.current = ctl;
//   }
//   return (
//     <FeatureGroup ref={editLayerRef}>
//       <EditControl position="topright" onMounted={onMounted} {...props} />
//     </FeatureGroup>
//   );
// }
// function EditableGroup(props) {
//   const [selectedLayerIndex, setSelectedLayerIndex] = useState(0);

//   function handleLayerClick(e, drawControl) {
//     setSelectedLayerIndex(e.target.feature.properties.editLayerId);
//   }

//   let dataLayer = new L.GeoJSON(props.data);
//   let layers = [];
//   let i = 0;
//   dataLayer.eachLayer((layer) => {
//     // Create a deep copy of the feature object
//     // let featureCopy = JSON.parse(JSON.stringify(layer.feature));
//     // Add the editLayerId property to the new object
//     layer.feature.properties.editLayerId = i;
//     // Create a new layer with the copied feature object
//     // let newLayer = L.geoJSON(featureCopy);
//     layers.push(layer);
//     i++;
//   });
//   console.log("hello", dataLayer);
//   return (
//     <div>
//       {layers.map((layer, i) => {
//         return (
//           <EditableLayer
//             key={i}
//             layer={layer}
//             showDrawControl={i === selectedLayerIndex}
//             onLayerClicked={handleLayerClick}
//           />
//         );
//       })}
//     </div>
//   );
// }
const googleAPIKey = process.env.NEXT_PUBLIC_GOOGLE_MAP_API_KEY?.toString();

const LeafletMap = ({ polygon, handlePolygon }: ISectorPolygon) => {
  const center: L.LatLngExpression = [45.76342, 4.834277];

  const render = (status: Status): React.ReactElement => {
    if (status === Status.LOADING) return <CommonSpinner />;
    return <p>Error</p>;
  };

  return (
    <>
      {googleAPIKey && (
        <Wrapper apiKey={googleAPIKey} render={render}>
          <MapContainer
            className="c-LeafletMap"
            center={center}
            zoom={10}
            maxZoom={30}
            scrollWheelZoom={false}
            //TODO: Add Bounds of france
            // maxBounds={franceBounds}
            maxBoundsViscosity={0.5}
            worldCopyJump={true}
            crs={CRS.EPSG3857}
          >
            <TileLayer
              url="https://{s}.google.com/vt/lyrs=m&x={x}&y={y}&z={z}"
              subdomains={["mt0", "mt1", "mt2", "mt3"]}
              attribution={"&copy; Google"}
            />

            <EditableGroup
              polygon={polygon}
              handlePolygon={handlePolygon}
              //TODO: Send Commune
              // commune={commune}
            />
          </MapContainer>
        </Wrapper>
      )}
      <div>Invalid google api key</div>
    </>
  );
};

export default LeafletMap;
