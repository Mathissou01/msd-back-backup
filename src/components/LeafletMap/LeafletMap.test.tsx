import { render, screen } from "@testing-library/react";
import LeafletMap from "./LeafletMap";

it("renders", () => {
  const { container } = render(<LeafletMap />);
  expect(container).toMatchSnapshot();
});

it("has map container", () => {
  render(<LeafletMap />);
  // MapContainer from react-leaflet is mocked and only returns <div/>
  const mapContainer = screen.getByTestId("map-container");
  expect(mapContainer).toBeInTheDocument();
});

it("has a result button", () => {
  render(<LeafletMap />);
  const button = screen.getByRole("button");
  expect(button).toBeInTheDocument();
});
