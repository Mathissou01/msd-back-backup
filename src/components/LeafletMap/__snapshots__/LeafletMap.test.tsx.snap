// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div>
  <div
    class="c-LeafletMap"
  >
    <div
      data-testid="map-container"
    />
    <button
      class="c-LeafletMap__Button"
    >
      Récupérer Adresses
    </button>
    ;
    <div
      class="c-LeafletMap__Results"
    >
      Résultats (
      0
      ) :
    </div>
    ;
  </div>
</div>
`;
